samples/durable-task-sdks/dotnet/Agents/PromptChaining/Worker/Services/BaseAgentService.cs

```diff
@@ -96,99 +71,59 @@ protected BaseAgentService(string endpointUrl, ILogger<BaseAgentService> logger,
         };
     }
     
-
-    
     /// <summary>
-    /// Ensures the agent exists, creating it if necessary
+    /// Initializes the agent if needed, ensuring it's created with the system prompt
+    /// Thread-safe implementation for concurrent access
     /// </summary>
-    public async Task<string> EnsureAgentExistsAsync(string agentName, string systemPrompt)
+    protected async Task InitializeAsync()
     {
-        Logger.LogInformation($"Setting up agent: {agentName}");
-        Logger.LogInformation($"Using Azure AI Projects endpoint: {Endpoint}");
-
-        try
-        {
-            // Check if an agent with this name already exists
-            PersistentAgent? existingAgent = null;
-            string existingAgentId = string.Empty;
+        // First fast check without a lock (optimistic)
+        if (_initialized)
+            return;
             
-            try
-            {
-                Logger.LogInformation($"Attempting to retrieve agents from Azure AI Projects endpoint: {Endpoint}");
-                var agents = AgentsClient.Administration.GetAgents();
-                
-                // Try to find an existing agent with the same name
-                foreach (var agent in agents)
-                {
-                    if (agent.Name == agentName)
-                    {
-                        existingAgent = agent;
-                        existingAgentId = agent.Id;
-                        Logger.LogInformation($"Found existing agent: {agentName} with ID: {existingAgentId}");
-                        break;
-                    }
-                }
-            }
-            catch (RequestFailedException rfEx)
+        bool shouldInitialize = false;
+            
+        // Use lock only to check/update the flag
+        lock (_initializationLock)
+        {
+            // Second check inside the lock (pessimistic)
+            if (!_initialized)
             {
-                Logger.LogWarning($"Error getting agents: {rfEx.Message}. Status: {rfEx.Status}");
-                throw; // Re-throw the exception to surface it
+                // Mark that this thread will do the initialization
+                shouldInitialize = true;
             }
-            
-            string agentId;
-            
-            if (existingAgent == null)
+        }
+        
+        // If we need to initialize, do it outside the lock to avoid blocking other threads
+        if (shouldInitialize)
```

Comment:
I don't think we can use this locking approach as we don't just need one thread perform initialization, we need *all other* threads to wait until it's done.  For that, we'd probably want a `SemaphoreSlim` that allows just a single thread in to do initialization but keeps the other threads waiting.